### PR TITLE
Orca bigdl estimator

### PR DESCRIPTION
**Pull request by cyita**

### CODE REVIEW

**Comment by cyita** [comment]
Jenkins passed http://10.239.47.210:18888/job/ZOO-PR-Validation/4303/

File: pyzoo/zoo/orca/learn/utils.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+def find_latest_checkpoint(model_dir, model_type="bigdl"):
+    import os
+    import re
+    import datetime
+    ckpt_path = None
+    latest_version = None
+    optim_prefix = None
+    optim_regex = None
+    if model_type == "bigdl":
+        optim_regex = "Sequential[0-9a-z]{8}\.[0-9]+$"
```
Comment (qiuxin2012) [inline pyzoo/zoo/orca/learn/utils.py:27]: not all bigdl model is contained in Sequential。。。

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+        if model_dir is not None:
+            self.model_dir = model_dir
+
+        if is_checkpoint:
+            from zoo.orca.learn.utils import find_latest_checkpoint
+            from zoo.pipeline.api.net import Net
+            from bigdl.nn.layer import Model
+            from bigdl.optim.optimizer import OptimMethod
+            import os
+            path, prefix, version = find_latest_checkpoint(checkpoint, model_type="bigdl")
```
Comment (qiuxin2012) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:316]: always load bigdl？

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+#
+from zoo.pipeline.nnframes import NNEstimator
+from zoo.pipeline.estimator import Estimator as SparkEstimator
+from zoo.orca.data import SparkXShards
+from bigdl.optim.optimizer import MaxEpoch
+from zoo.feature.common import FeatureSet
+from pyspark.sql.dataframe import DataFrame
+
+
+class Estimator(object):
```
Comment (qiuxin2012) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:24]: Why we need this Estimator?

Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:24]: Because this estimator is an interface, and there may be new wrappers added to the BigDL estimator.

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+        pass
+
+    def get_train_summary(self):
+        pass
+
+    def get_validation_summary(self):
+        pass
+
+    @staticmethod
+    def from_bigdl(*, model, loss, optimizer=None, feature_preprocessing=None,
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:74]: loss should be optional too

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+        self.feature_preprocessing = feature_preprocessing
+        self.label_preprocessing = label_preprocessing
+        self.model_dir = model_dir
+        self.model = model
+        self.nn_model = None
+        self.estimator = None
+        self.log_dir = None
+        self.app_name = None
+
+    def fit(self, data, epochs, feature_cols="features", optimizer=None, batch_size=32,
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:124]: is it consistent with, say, tf.Estimator.fit? Seems that `feature_cols` is missing

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+    def __init__(self, *, model, loss, optimizer=None, feature_preprocessing=None,
+                 label_preprocessing=None, model_dir=None):
+        self.loss = loss
+        self.optimizer = optimizer
+        self.feature_preprocessing = feature_preprocessing
+        self.label_preprocessing = label_preprocessing
+        self.model_dir = model_dir
+        self.model = model
+        self.nn_model = None
+        self.estimator = None
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:120]: I think you should also have an `nn-estimator` instead of reusing one estimator

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 5 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:128]: `feature_cols` exists in tf.Estimator.fit too. Do you mean `labels_cols`?

Comment (jason-dai) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:132]: yes, `labels_cols`

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 7]
```diff
@@ -116,10 +116,18 @@ def __init__(self, *, model, loss, optimizer=None, feature_preprocessing=None,
...
-        self.estimator = None
+        self.nn_model = NNModel(self.model, feature_preprocessing=self.feature_preprocessing)
+        self.nn_estimator = NNEstimator(self.model, self.loss, self.feature_preprocessing,
+                                        self.label_preprocessing)
+        if self.optimizer is not None:
+            self.nn_estimator.setOptimMethod(self.optimizer)
+        if self.optimizer is None:
+            from bigdl.optim.optimizer import SGD
+            optimizer = SGD()
+        self.estimator = SparkEstimator(self.model, optimizer, self.model_dir)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:127]: Are there big overheads associated with creating `NNEstimator` or `SparkEstimator`?

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 5 above)
Comment (jason-dai) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:132]: And why have the user pass `optimizer` in `fit` again?

Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:132]: Users may want to pass a data-related optimizer like https://github.com/dell-ai-engineering/BigDL-ImageProcessing-Examples/blob/master/ChestXray-Training.ipynb `get_adam_optimMethod`.

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 7 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:127]: It takes 6-7ms to create an Estimator/NNEstimator.

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 5 above)
Comment (jason-dai) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:132]: But that can still be specified when creating the Estimator (e.g., when creating NNEstimator in https://github.com/dell-ai-engineering/BigDL-ImageProcessing-Examples/blob/master/ChestXray-Training.ipynb), no?

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 7 above)
Comment (jason-dai) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:127]: Is it Spark local or YARN?

Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:127]: Spark local.

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 5 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:132]: Yes, I'll remove the optimizer param in fit.

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 7 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:127]: DistriOptimizer will be created when calling `fit` in `Estimator` or `NNEstimator`. And the model won't be cached until the fit function was called.

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        self.optimizer = optimizer
+        self.feature_preprocessing = feature_preprocessing
+        self.label_preprocessing = label_preprocessing
+        self.model_dir = model_dir
+        self.model = model
+        self.nn_model = NNModel(self.model, feature_preprocessing=self.feature_preprocessing)
+        self.nn_estimator = NNEstimator(self.model, self.loss, self.feature_preprocessing,
+                                        self.label_preprocessing)
+        if self.optimizer is not None:
+            self.nn_estimator.setOptimMethod(self.optimizer)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:123]: need to move `setOptimMethod` after `optimizer = SGD()`?

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+            raise NotImplementedError
+        elif isinstance(data, SparkXShards):
+            from zoo.orca.data.utils import to_sample
+            from zoo.orca.learn.metrics import Metrics
+
+            validation_methods = Metrics.convert_metrics_list(validation_methods)
+            val_feature_set = FeatureSet.sample_rdd(data.rdd.flatMap(to_sample))
+            return self.estimator.evaluate(val_feature_set, validation_methods, batch_size)
+        else:
+            raise ValueError("Data should be SparkXShards or Spark DataFrame, but get " +
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:258]: `Data should be XShards, but ...`

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+                    feature_cols = feature_cols[0]
+                else:
+                    from pyspark.ml.feature import VectorAssembler
+                    assembler = VectorAssembler(
+                        inputCols=feature_cols,
+                        outputCol="features")
+                    data = assembler.transform(data)
+                    if val_data is not None:
+                        val_data = assembler.transform(val_data)
+                    feature_cols = "features"
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:152]: should refact these logic into a utility function and re-used in multiple places

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 8 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:123]: NNEstimator can fit without optimizer. If the two estimators should have the same behavior, `setOptimMethod` should be moved after `optimizer = SGD()`.

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+                train_summary = TrainSummary(log_dir=self.log_dir, app_name=self.app_name)
+                self.nn_estimator.setTrainSummary(train_summary)
+                val_summary = ValidationSummary(log_dir=self.log_dir, app_name=self.log_dir)
+                self.nn_estimator.setValidationSummary(val_summary)
+            if self.model_dir is not None and checkpoint_trigger is not None:
+                checkpoint_trigger = Trigger.convert_trigger(checkpoint_trigger)
+                self.nn_estimator.setCheckpoint(self.model_dir, checkpoint_trigger)
+
+            self.nn_model = self.nn_estimator.fit(data)
+            self.is_last_fit_nn = True
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:190]: is_nnframe_fit

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+            val_methods = Metrics.convert_metrics_list(val_methods)
+            checkpoint_trigger = Trigger.convert_trigger(checkpoint_trigger)
+
+            if isinstance(data, SparkXShards):
+                train_rdd = data.rdd.flatMap(to_sample)
+                train_feature_set = FeatureSet.sample_rdd(train_rdd)
+                if val_data is None:
+                    val_feature_set = None
+                else:
+                    assert isinstance(val_data, SparkXShards), "val_data should be a SparkXShards"
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:204]: val_data should be a XShards

File: pyzoo/zoo/orca/learn/bigdl/estimator.py [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+                self.nn_model.setSamplePreprocessing(sample_preprocessing)
+            return self.nn_model.transform(data)
+        elif isinstance(data, SparkXShards):
+            from zoo.orca.data.utils import to_sample
+            from zoo.orca.learn.tf.utils import convert_predict_to_xshard
+            sample_rdd = data.rdd.flatMap(to_sample)
+            result_rdd = self.model.predict(sample_rdd)
+            return convert_predict_to_xshard(result_rdd)
+        else:
+            raise ValueError("Data should be SparkXShards or Spark DataFrame, but get " +
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/bigdl/estimator.py:242]: Data should be XShards or Spark DataFrame

File: pyzoo/zoo/orca/learn/bigdl/estimator.py (hunk 8 above)
Comment (jason-dai) [reply pyzoo/zoo/orca/learn/bigdl/estimator.py:123]: Yes, let's make them have the same behavior

**Comment by cyita** [comment]
Jenkins passed http://10.239.47.210:18888/job/ZOO-PR-Validation/4440/

**Comment by jason-dai** [comment]
LGTM
